hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/ValidationSupportChain.java

```diff
@@ -393,56 +803,520 @@ public CodeValidationResult validateCodeInValueSet(
+
+		LookupCodeKey key = new LookupCodeKey(theLookupCodeRequest);
+		CacheValue<LookupCodeResult> retVal = getFromCache(key);
+		if (retVal == null) {
+
+			retVal = CacheValue.empty();
+			for (IValidationSupport next : myChain) {
+				final String system = theLookupCodeRequest.getSystem();
+				final String code = theLookupCodeRequest.getCode();
+				final String displayLanguage = theLookupCodeRequest.getDisplayLanguage();
+				if (isCodeSystemSupported(theValidationSupportContext, next, system)) {
+					LookupCodeResult lookupCodeResult =
+							next.lookupCode(theValidationSupportContext, theLookupCodeRequest);
+					if (lookupCodeResult == null) {
+						/*
+						This branch has been added as a fall-back mechanism for supporting lookupCode
+						methods marked as deprecated in interface IValidationSupport.
+						*/
+						//noinspection deprecation
+						lookupCodeResult = next.lookupCode(theValidationSupportContext, system, code, displayLanguage);
+					}
+					if (lookupCodeResult != null) {
+						ourLog.debug(
+								"Code {}|{}{} {} by {}",
+								system,
+								code,
+								isBlank(displayLanguage) ? "" : " (" + theLookupCodeRequest.getDisplayLanguage() + ")",
+								lookupCodeResult.isFound() ? "found" : "not found",
+								next.getName());
+						retVal = new CacheValue<>(lookupCodeResult);
+						break;
+					}
 				}
-				return lookupCodeResult;
 			}
+
+			putInCache(key, retVal);
+		}
+
+		return retVal.getValue();
+	}
+
+	/**
+	 * Returns a view of the {@link IValidationSupport} modules within
+	 * this chain. The returned collection is unmodifiable and will reflect
+	 * changes to the underlying list.
+	 *
+	 * @since 8.0.0
+	 */
+	public List<IValidationSupport> getValidationSupports() {
+		return Collections.unmodifiableList(myChain);
+	}
+
+	private <T> void putInCache(BaseKey<T> key, CacheValue<T> theValue) {
+		if (myExpiringCache != null) {
+			myExpiringCache.put(key, theValue);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> CacheValue<T> getFromCache(BaseKey<T> key) {
+		if (myExpiringCache != null) {
+			return (CacheValue<T>) myExpiringCache.getIfPresent(key);
+		} else {
+			return null;
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private List<IBaseResource> getFromCacheWithAsyncRefresh(
+			FetchAllKey theKey, Supplier<List<IBaseResource>> theLoader) {
+		if (myExpiringCache == null || myNonExpiringCache == null) {
+			return theLoader.get();
+		}
+
+		CacheValue<List<IBaseResource>> retVal = getFromCache(theKey);
+		if (retVal == null) {
+			retVal = (CacheValue<List<IBaseResource>>) myNonExpiringCache.get(theKey);
+			if (retVal != null) {
+				Runnable loaderTask = () -> {
+					List<IBaseResource> loadedItem = theLoader.get();
+					CacheValue<List<IBaseResource>> value = new CacheValue<>(loadedItem);
+					myNonExpiringCache.put(theKey, value);
+					putInCache(theKey, value);
+				};
+				myBackgroundExecutor.execute(loaderTask);
+
+				return retVal.getValue();
```

Comment:
Is there a race condition here? If so, consider guarding with

```suggestion
				returnValue = retVal.getValue();
				
				myBackgroundExecutor.execute(loaderTask);

				return returnValue;
```